### PR TITLE
Add 'threadable' class behaviour modifier, similar to ISmaccUpdatable

### PR DESCRIPTION
**Pull request by yassiezar**
This PR adds a class that can be used to modify the behaviour of SMACC entities, allowing them to offload long-running processes onto a threadpool attached to the entity's lifetime. The goal is to be similar to the current ISmaccUpdatable modifier, which adds additional functionality to CB's clients and components. 

The use-case for this functionality would be in cases where you have long-running processes being executed within ROS callbacks, which can block other callbacks from being triggered and processed. In our case, this caused teleoperation signals to be executed with significant delay - an obvious safety issue. 

Extending a SMACC entity with this modifier gives it access to a threadpool with a set number of threads (default = 1), as well as a job queue of fixed size (default = 1). The client then pushes a callback object containing the long-running process (functor, std::function, lambda, etc.) with the arguments taken from the ROS callback input onto the job queue where a processor thread is assigned to execute it. The job queue has a fixed size with a circular buffer, meaning that only the `n` most recent jobs are held for processing (n=size of queue), discarding older jobs as new ones come in. Calling `stop()` will terminate the threadpool and will typically be done in `onExit()` calls. I'm wondering if there's a nice way to automatically do this when the parent entity goes out of scope

I've included a small example to showcase the functionality and demonstrate how it can be implemented. Hopefully others find this useful and I'd be happy to hear any suggestions for improvement

┆Issue is synchronized with this [Jira Task](https://robosoft-ai.atlassian.net/browse/SMACC1-10) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by pabloinigoblasco** [comment]
Hello.
Your proposal is a bit disorienting to me :-) , but at the same time looks to me quite interesting.

I am still trying to understand how this is related with the process of designing and implementing state machines with SMACC and how this tool is associated to their concepts (clients, states, client behaviors, lifetimes=

At first sight it looks to me like a "c++ library or tool" that could be a class used as an attribute of your smacc entity (using composition instead of inheritance). Making it a modifier (using inheritance) or a new type of smacc entitiy (such as clients, clientbehaviors, components) looks to me initially a bit intrusive.

But you also mentioned something interesting, you would like to associate that "thread pool" to the lifetime of some short-time object. 

The first I wonder is which use cases are you thinking on to need this tool (I think it could be useful but I am not being able to imagine any use case not covered by our actual system - I do not discuss that though).

We have essentially two types short-life-time objects: States, state reactors and client behaviors. Would this be a new type? I see something interesting here. Let me share your same idea but from my own point of view:

I see three designs about how a thread pool utility like this could be used to fit better SMACC without creating a new smacc entity:

1 - The threadpool could be essentially a client. That is, a long-life object that is always alive (even if it is idle).
The callbacks or jobs associated to some state would be configured as client behaviors. CbJob::OnExit would unregister/cancel the callback from the thread pool.

2- If you actually need a short-life threapool I would say that the threadpool is just another asynchronous client behavior (because of its lifetyme). The only missing functionality is the ability of use that short-life threadpool from other clientbehaviors. In order to achieve that goal I would implement a requiresClientBehavior functionality to create dependencies and automatic binding between client behaviors.

3 - At the current moment Thread parallelism in states is covered by asynchronous client behaviors (short-lifetime objects). But it is true that we open one thread for each one and that could be in some cases expensive (specially for short jobs). Would be interesting to have a thread pool for asynchronous client behaviors? Maybe yes.

---
In conclusion, I think it is always nice to have tools for SMACC. However I have two concerns:
 - first I don't see clearly which cases this would cover that is not actually covered by asynchronous client behaviors. 
 - I think the two exposed models fits better the current SMACC model

What do you think about this?

**Comment by yassiezar** [comment]
Hey @pabloinigoblasco, thanks for having a look and the detailed response.

The use-case for a thread pool like this would be to provide subscriber CBs with an additional set of threads to perform some long-running processes off of the main ROS spinner thread. For example, we have a CB that defines a callback that is triggered on every input camera frame and performs some expensive image processing. If we leave the image segmentation to run on the thread the callback is triggered on (i.e. the ROS spinner thread), then it prevents other inputs from being processed elsewhere in the state machine (e.g. teleop signals) until the image has finished being processes and the callback is returned from.

The question of 'does this functionality aid SMACC's core goals' is fair. However, the existence of the current async CBs demonstrate that parallelism is already an important consideration within SMACC's design. Normally I'd agree with you and use the threadpool as a standalone class through composition. However, I propose in this PR that we use it as a modifier instead similar to the existing 'updatable' modifier base class. The advantages are that

1. it follows SMACC's existing style of defining and extending class behaviour through inheritance, 
2. it allows for strong control of the pool's lifetime behaviour (e.g. the base class can ensure that the pool follows some generic shutdown behaviour which can be extended through the inheritor's `onEntry` and `onExit` methods) which might not be as easily controlled within a class destructor, and
3. it can be used to extend the behavior of any base SMACC entity to provide it with a threadpool with which to perform expensive computation (e.g. allow SMACC's existing pose component to monitor the robot pose on a separate thread, make an updatable client that performs potentially blocking requests, a CB that performs expensive image processing, etc.)

Note that this class will *not* be a 'new' SMACC entity, but rather will allow users to extend existing entities with new behaviours *a la* the `Updatable` base class. I.e. by itself, it won't do anything and needs to exist alongside another SMACC entity (client, CB, component)

SMACC's existing async CBs are great and solved some of the initial issues we had regarding expensive CB setup and teardowns. However, if we wanted to extend that functionality beyond async `onEntry` and `onExit` calls (which I think they're currently limited to last I checked?), then we'd need to create an async CB for topic subscribers, publishers and service clients and servers and do the same for clients and components (lots of repetition!). Instead, we can provide users with tools to extend the existing clients and CBs to provide them with async functionality by simly inheriting from the `Threadable` class. Unfortuneately, threads are quite expensive to spin up so they might not be ideal for short-lived objects like CBs where `std::async` is arguably a better fit, but maybe the convenience, ease of use and flexibility is worth the trade-off? 

Regarding your 3 proposals:

1. This is an interesting idea, I like it! My only worry is that SMACC currently seems to use clients as an interface dedicated to a specific topic, service or some hardware entity and this might not be in keeping with that spirit?

2. Also an intersting idea and will reduce the repetition I mentioned above, but it sounds like it would need a significant rewrite within the SMACC library to add that CB dependency? I also don't know if its a good idea to create those CB interdependencies. 

3. I think the current sync CB implementation is fine as-is, but it can perhaps be made cleared that the async bits are limited to the `onEntry` and `onExit` methods. `std::async` is a much cheaper than `std::thread`, since it uses its own thread pool under the hood so its a better fit for short-lived CBs than a manual threadpool is. However, see my above point about this trade-off. 

> I think the two exposed models fits better the current SMACC mode

Could you elaborate? I'm not sure I follow which 2 models you are refering to.

In short, I think that SMACC's existing parallelism could be extended to include long-running processes running outside CB `onEntry` and `onExit` calls, including clients and maybe components. I've tried to make it as simple as possible to implement such behaviour in existing SMACC entities without modifying the base library. Of course, I leave it to you and the community to decide exactly how such functionality is implemented, but I hope I showed the benefits of this functionality where SMACC is used and the use-cases where it might be beneficial

**Comment by pabloinigoblasco** [comment]
Hello.
I am having another pass to this to understand your goal with the new example

**Comment by pabloinigoblasco** [comment]
Hello @yassiezar I have been studying your use case.
It is interesting because it shows "a challenge" we did not have to face in any of our other use cases. 

The solution you propose is concise and effective, however, I believe it introduces a new core concept to solve a problem that already has a solution in ROS. 

The issue of subscriber callbacks needing to be fast to prevent blocking the main loop and delaying other callbacks is a common problem in ROS. 

ROS already offers a solution for this: the ros::AsyncSpinner) concept 
For example, refer to this link: https://roboticsbackend.com/ros-asyncspinner-example/. 

I attempted to modify the signal_detector code to incorporate this type of spinner and the results appear to be the same."

```
$ git diff smacc
diff --git a/smacc/src/smacc/signal_detector.cpp b/smacc/src/smacc/signal_detector.cpp
index 57b0177d..f543d854 100644
--- a/smacc/src/smacc/signal_detector.cpp
+++ b/smacc/src/smacc/signal_detector.cpp
@@ -249,6 +249,9 @@ void SignalDetector::pollingLoop()
 
   ros::NodeHandle _;
   ros::Rate r0(20);
+  ros::AsyncSpinner asyncSpinner(0);
+  asyncSpinner.start();
+
   while (!initialized_)
   {
     r0.sleep();
@@ -276,8 +279,8 @@ void SignalDetector::pollingLoop()
   {
     ROS_INFO_STREAM_THROTTLE(10, "[SignalDetector] heartbeat");
     pollOnce();
-    ros::spinOnce();
     r.sleep();
   }
+  ros::waitForShutdown();
 }
 }  // namespace smacc
 
 ```

Results with the non-threaded client behaviors are the same than the threaded one:

```
[ INFO] [1673992699.801111825]: HIGH frequency out
[ INFO] [1673992699.867648164]: HIGH frequency out
[ INFO] [1673992699.934243973]: HIGH frequency out
[ INFO] [1673992700.000887466]: HIGH frequency out
[ INFO] [1673992700.067578080]: HIGH frequency out
[ INFO] [1673992700.134215787]: HIGH frequency out
[ INFO] [1673992700.200874818]: HIGH frequency out
[ INFO] [1673992700.201179429]: LOW frequency out
[ INFO] [1673992700.267623677]: HIGH frequency out
[ INFO] [1673992700.334181201]: HIGH frequency out
[ INFO] [1673992700.400863471]: HIGH frequency out
[ INFO] [1673992700.467553138]: HIGH frequency out
[ INFO] [1673992700.534149043]: HIGH frequency out
[ INFO] [1673992700.600864718]: HIGH frequency out
[ INFO] [1673992700.601318636]: LOW frequency out
```

At this point, unless new information arises, I believe that incorporating support for the asynchronous spinner mode for the signal detector would address this use case effectively.

Does this solution fully address all the use cases you had in mind? Additionally, should we consider implementing a thread pool component? I am not certain at this time, but I am open to further discussion and understanding your reasoning.

What are your thoughts on this?

**Comment by yassiezar** [comment]
@pabloinigoblasco this is a great result and it makes sense to use ROS's built-in tools to enable multithreading and I'm perfectly happy to use this approach. My only concern might be whether there may be any knock-on effects and consequences on others' projects if we only use AsyncSpinners? Unless you were thinking of adding an option to select a single or multithreaded spinner

I'm thinking that a threadpool component would still be useful if tighter control over the job execution is needed, but its not a major concern for our use-case at the moment

**Comment by pabloinigoblasco** [comment]
We are currently working on implementing the AsyncSpinner funcionality in a backwards-compatible manner. Your suggestion of using it optionally is a great idea. By default, SMACC will run in single-threaded mode in its "ROS loop" like other ROS applications, but it will also have the option to run in async mode. 

Please note that this pull request has not been merged, some parts such as the state machine example might be interesting to be merged. Regardless, your contributions have been invaluable and have helped the project greatly. Thank you.

**Comment by yassiezar** [comment]
@pabloinigoblasco great, glad to hear you've found it helpful. How would you like to proceed and what do you need from me?

**Comment by pabloinigoblasco** [comment]
I think we are done with the support of this funcionality. It is available in melodic and noetic. (I am also porting that to smacc2)

Have a look to this function:
https://github.com/robosoft-ai/SMACC/blob/4331940880b50d8a0892cea7b33fe3b6ed810acf/smacc/include/smacc/smacc_signal_detector.h#L88

I would suggest you to test your example with this functionality to see if your  needs are really solved.
Thanks again for your help.

**Comment by yassiezar** [comment]
@pabloinigoblasco thanks for implementing these. We'll be sure to take a look at it and let you know how it works out. I'm going to keep the threadpool and modify it as a component to try and address another use-case I came across (non-ROS managed file read/writes). I'll re-open a PR when I'm happy with it, so I'll go ahead and close this one
